crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+                            // we can assume that for any error
+                            // we need to retry all the inflight bathes.
+                            // special case for load shedding we could
+                            // throttle the stream a little bit then
+                            // speed up over a period of time.
+
+                            break SessionState::Disconnected;
+                        }
+                    }
+                }
+                _ = ctx.cancelled() => {
+                    // relies on auto drain and drop of the inflight
+                    // batches to notify callers that records has been
+                    // cancelled.
+                    return SessionState::Shutdown;
+                }
+            }
+        };
+
+        // don't lose the buffered batch in case of ChunksTimeout
+        if let Either::Left(chunked) = chunked.as_pin_mut() {
+            let remainder = chunked.into_remainder();
+            if !remainder.is_empty() {
+                let (trackers, records): (Vec<_>, Vec<_>) = remainder.into_iter().unzip();
+                let batch: Arc<[IngestRecord]> = Arc::from(records);
+                self.inflight.push_back(IngressBatch {
+                    records: batch,
+                    trackers,
+                    reply_rx: None,
+                });
+            }
+        }
+
+        state
+    }
+}
+
+struct SessionManagerInner<T> {
+    networking: Networking<T>,
+    partition_routing: PartitionRouting,
+    opts: SessionOptions,
+    ctx: CancellationToken,
+    published: ArcSwap<HashMap<PartitionId, SessionHandle>>,
+    locks: DashMap<PartitionId, OnceLock<SessionHandle>>,
+}
+
+impl<T> SessionManagerInner<T>
+where
+    T: TransportConnect,
+{
+    /// Gets or start a new session to partition with given partition id.
+    /// It guarantees that only one session is started per partition id.
+    pub fn get(&self, id: PartitionId) -> SessionHandle {
+        let inner = self.published.load();
+        match inner.get(&id) {
+            Some(handle) => handle.clone(),
+            None => {
+                let once = self.locks.entry(id).or_default();
+
+                let handle = once.get_or_init(|| {
+                    let session = PartitionSession::new(
+                        self.networking.clone(),
+                        self.partition_routing.clone(),
+                        id,
+                        self.opts.clone(),
+                    );
+
+                    let handle = session.handle();
+
+                    //todo(azmy): handle spawn result
+                    let ctx = self.ctx.clone();
+                    let _ = TaskCenter::spawn_child(
+                        TaskKind::Background,
+                        "ingress-partition-session",
+                        async move {
+                            session.start(ctx).await;
+                            Ok(())
+                        },
```

Comment:
When spawning a child, you don't have to manually pass in `cancellation_token().child_token()`. This happens automatically if you call `cancellation_token()` from within the child task.

Reply:
The reason i still pass a cancellation token is that ingress client can be shared among different components. Partition Sessions are only started on demand, hence they can become "child task" for different parents.

To avoid a parent task shutting down from killing partition session that might be used by others was to make sure they only share the same cancellation token that is owned by the client itself.

Maybe `spawn_child` then was not the best spawn function to use?

Reply:
Ah ok. So far it seems that we are taking the `CancellationToken` from the `SessionManager` when it is being created.